add the architecture doc (#144)
protofire/solhint: docs/architecture.md

---
layout:      default
title:       Solhint Project Structure
date:        2019-08-27 16:11:00 +0000
author:      "@think-in-universe"
description: Introduction of the structure of the project, how it works, and how to add new rules, etc.
---

# Architecture

> This document is written for the developers who want to understand the structure and mechanism of this project, and how to contribute.

Reviewer: ```suggestion
This document is for developers who want to understand the structure and mechanisms of this project, and how to contribute.
```

## Project Structure

> Below describes the major structure of the project.

Reviewer: ```suggestion
The file structure below describes the major structure of the project.
```

```
├── docs                                 # documentation
├── lib                                  # main source code
│   ├── comment-directive-parser.js      # comment parsers
│   ├── common                           # utility modules for syntax parsing, reporting, etc.
│   ├── config                           # helpers for loading solhint configuration
│   ├── config.js                        # load config
│   ├── doc                              # documentation utilities
│   ├── grammar                          # solidity grammar, generated with ANTLR
│   ├── index.js                         # main entry point
│   ├── load-rules.js                    # module for loading rules
│   ├── reporter.js                      # module for reporting results
│   ├── rules                            # source for solhint rules
│   └── tree-listener.js                 # used for register all loaded rules
├── scripts                              # script for generating grammar and rule docs
├── solhint.js                           # solhint command line
├── solidity-antlr4                      # git submodule for solidity-antlr4
└── test                                 # tests
```

> ## How it Works

Reviewer: ```suggestion
## How solhint Works
```

### Solidity Grammar with ANTLR

Solhint depends on [ANTLR4](http://www.antlr.org/) to generate the Solidity parser, following the grammar description taken from
[solidity-antlr4](https://github.com/solidityj/solidity-antlr4).

> To update the Solidity grammar, you'll need to update the Git submodule solidity-antlr4:

Reviewer: ```suggestion
To update the Solidity grammar, you need to update the Git submodule solidity-antlr4:
```

```sh
git submodule update --remote # update to latest version
# or checkout a specific commit or tag:
# cd solidity-antlr4
# git checkout abc123
```

> Then execute `scripts/build-grammar.sh`. (Java Runtime Environment 7, or later, is required for running the script.)

Reviewer: ```suggestion
Then run `scripts/build-grammar.sh`. (Java Runtime Environment 7, or later, is required for running the script.)
```

### How to Add A New Rule

> The Solhint rules in `lib/rules` contains the different aspects of lint requirements, such as naming, best practices, security, etc.

Reviewer: ```suggestion
The Solhint rules in `lib/rules` contains the different lint requirements, such as naming, best practices, security, etc.
```

> The rules are implemented with a visitor pattern. You can extend the `BaseChecker` class with the `ruleId` and `meta` fields to define a rule, and implement methods that are called when a node in the AST is entered or exited. The constructor accepts a reporter and a config, and `ruleId` field is present in the object. This `ruleId` is the one that will be used to activate and configure the rule.

Reviewer: ```suggestion
The rules are implemented with a [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern). You can extend the `BaseChecker` class with the `ruleId` and `meta` fields to define a rule, and implement methods that are called when a node in the AST is entered or exited. The constructor accepts a reporter and a config, and `ruleId` field is present in the object. This `ruleId` is the one that will be used to activate and configure the rule.
```

```javascript

const ruleId = 'indent'
const DEFAULT_SEVERITY = 'error'
const DEFAULT_INDENTS = 4
const meta = {
  type: 'align',

  docs: {
    // ...
  },

  isDefault: true,
  recommended: true,
  defaultSetup: [DEFAULT_SEVERITY, DEFAULT_INDENTS],

  schema: [
    {
      type: 'array',
      items: [{ type: 'integer' }],
      uniqueItems: true,
      minItems: 2
    }
  ]
}

class IndentChecker {
  constructor(reporter, config) {
    this.reporter = reporter
    this.ruleId = ruleId
    this.meta = meta
    this.linesWithError = []

    const indent = this.parseConfig(config).indent || 4
    const indentUnit = this.parseConfig(config).unit || 'spaces'

    this.blockValidator = new BlockValidator(indent, indentUnit, reporter, this.ruleId)

    // ...
  }

  enterBlock(ctx) {
    this.blockValidator.validateBlock(ctx)
  }

  enterContractDefinition(ctx) {
    this.blockValidator.validateBlock(ctx)
  }

  enterStructDefinition(ctx) {
    this.blockValidator.validateBlock(ctx)
  }

  enterEnumDefinition(ctx) {
    this.blockValidator.validateBlock(ctx)
  }

  enterImportDirective(ctx) {
    this.blockValidator.validateBlock(ctx)
  }

```

> The developers of the new rules need to have basic understanding about the concepts and structure of the AST, and execute the proper logic when certain nodes in the AST are visited.

Reviewer: ```suggestion
Developers of new rules need to have a basic understanding about the concepts and structure of the AST, and execute the proper logic when certain nodes in the AST are visited.
```
